File: helix-runtime/apps/ios/Sources/Tasks/TasksService.swift
import Foundation

class TasksService {
    private let gateway: GatewaySession

    init(gateway: GatewaySession) {
        self.gateway = gateway
    }

    // MARK: - Task Operations

    func getTasks(
        boardId: String,
        status: [TaskStatus]? = nil,
        limit: Int = 50,
        offset: Int = 0
    ) async throws -> [Task] {
        var params: [String: Any] = [
            "board_id": boardId,
            "limit": limit,
            "offset": offset,
        ]

        if let status = status {
            params["status"] = status.map { $0.rawValue }
        }

        let response = try await gateway.request("tasks.list", params: params)
        guard let data = response as? [[String: Any]] else {
            throw TaskError.decodingError("Expected array of tasks")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode([Task].self, from: jsonData)
    }

    func getTask(id: String) async throws -> Task {
        let response = try await gateway.request("tasks.get", params: ["id": id])
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected task object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(Task.self, from: jsonData)
    }

    func createTask(
        boardId: String,
        title: String,
        description: String? = nil,
        priority: TaskPriority = .medium,
        dueDate: Date? = nil,
        estimatedHours: Double? = nil,
        tags: [String] = []
    ) async throws -> Task {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601

        var params: [String: Any] = [
            "board_id": boardId,
            "title": title,
            "priority": priority.rawValue,
            "tags": tags,
        ]

        if let description = description {
            params["description"] = description
        }
        if let dueDate = dueDate {
            params["due_date"] = ISO8601DateFormatter().string(from: dueDate)
        }
        if let estimatedHours = estimatedHours {
            params["estimated_hours"] = estimatedHours
        }

        let response = try await gateway.request("tasks.create", params: params)
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected task object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(Task.self, from: jsonData)
    }

    func updateTask(
        id: String,
        title: String? = nil,
        description: String? = nil,
        status: TaskStatus? = nil,
        priority: TaskPriority? = nil,
        dueDate: Date? = nil,
        estimatedHours: Double? = nil,
        tags: [String]? = nil
    ) async throws -> Task {
        var params: [String: Any] = ["id": id]

        if let title = title { params["title"] = title }
        if let description = description { params["description"] = description }
        if let status = status { params["status"] = status.rawValue }
        if let priority = priority { params["priority"] = priority.rawValue }
        if let dueDate = dueDate {
            params["due_date"] = ISO8601DateFormatter().string(from: dueDate)
        }
        if let estimatedHours = estimatedHours { params["estimated_hours"] = estimatedHours }
        if let tags = tags { params["tags"] = tags }

        let response = try await gateway.request("tasks.update", params: params)
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected task object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(Task.self, from: jsonData)
    }

    func deleteTask(id: String) async throws {
        _ = try await gateway.request("tasks.delete", params: ["id": id])
    }

    // MARK: - Subtask Operations

    func getSubtasks(taskId: String) async throws -> [Subtask] {
        let response = try await gateway.request("tasks.subtasks.list", params: ["task_id": taskId])
        guard let data = response as? [[String: Any]] else {
            throw TaskError.decodingError("Expected array of subtasks")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode([Subtask].self, from: jsonData)
    }

    func createSubtask(taskId: String, title: String) async throws -> Subtask {
        let response = try await gateway.request(
            "tasks.subtasks.create",
            params: ["task_id": taskId, "title": title]
        )
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected subtask object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(Subtask.self, from: jsonData)
    }

    func toggleSubtask(id: String, isCompleted: Bool) async throws -> Subtask {
        let response = try await gateway.request(
            "tasks.subtasks.toggle",
            params: ["id": id, "is_completed": isCompleted]
        )
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected subtask object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(Subtask.self, from: jsonData)
    }

    // MARK: - Board Operations

    func getBoards() async throws -> [KanbanBoard] {
        let response = try await gateway.request("tasks.boards.list", params: [:])
        guard let data = response as? [[String: Any]] else {
            throw TaskError.decodingError("Expected array of boards")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode([KanbanBoard].self, from: jsonData)
    }

    func getBoard(id: String) async throws -> KanbanBoard {
        let response = try await gateway.request("tasks.boards.get", params: ["id": id])
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected board object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(KanbanBoard.self, from: jsonData)
    }

    // MARK: - Search

    func searchTasks(filter: TaskSearchFilter, limit: Int = 50) async throws -> [Task] {
        var params: [String: Any] = ["limit": limit]

        if !filter.query.isEmpty { params["query"] = filter.query }
        if !filter.status.isEmpty { params["status"] = filter.status.map { $0.rawValue } }
        if !filter.priority.isEmpty { params["priority"] = filter.priority.map { $0.rawValue } }
        if let assigneeId = filter.assigneeId { params["assignee_id"] = assigneeId }
        if let boardId = filter.boardId { params["board_id"] = boardId }
        if !filter.tags.isEmpty { params["tags"] = filter.tags }
        if let hasSubtasks = filter.hasSubtasks { params["has_subtasks"] = hasSubtasks }
        if let isOverdue = filter.isOverdue { params["is_overdue"] = isOverdue }

        let response = try await gateway.request("tasks.search", params: params)
        guard let data = response as? [[String: Any]] else {
            throw TaskError.decodingError("Expected array of tasks")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode([Task].self, from: jsonData)
    }

    // MARK: - Analytics

    func getAnalytics(boardId: String) async throws -> TaskAnalytics {
        let response = try await gateway.request("tasks.analytics", params: ["board_id": boardId])
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected analytics object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(TaskAnalytics.self, from: jsonData)
    }

    // MARK: - Time Tracking

    func startTimeEntry(taskId: String) async throws -> TimeEntry {
        let response = try await gateway.request(
            "tasks.timetracking.start",
            params: ["task_id": taskId]
        )
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected time entry object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(TimeEntry.self, from: jsonData)
    }

    func stopTimeEntry(taskId: String) async throws -> TimeEntry {
        let response = try await gateway.request(
            "tasks.timetracking.stop",
            params: ["task_id": taskId]
        )
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected time entry object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(TimeEntry.self, from: jsonData)
    }

    func getTimeTrackingStats() async throws -> TimeTrackingStats {
        let response = try await gateway.request("tasks.timetracking.stats", params: [:])
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected stats object")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(TimeTrackingStats.self, from: jsonData)
    }

    // MARK: - Focus Time

    func getFocusTimeSettings() async throws -> FocusTime {
        let response = try await gateway.request("tasks.focustime.get", params: [:])
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected focus time object")
        }

        let decoder = JSONDecoder()
        let jsonData = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(FocusTime.self, from: jsonData)
    }

    func updateFocusTimeSettings(_ settings: FocusTime) async throws -> FocusTime {
        let encoder = JSONEncoder()
        let jsonData = try encoder.encode(settings)
        guard let params = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
            throw TaskError.invalidTaskData
        }

        let response = try await gateway.request("tasks.focustime.update", params: params)
        guard let data = response as? [String: Any] else {
            throw TaskError.decodingError("Expected focus time object")
        }

        let decoder = JSONDecoder()
        let jsonData2 = try JSONSerialization.data(withJSONObject: data)
        return try decoder.decode(FocusTime.self, from: jsonData2)
    }
}

File: helix-runtime/apps/android/app/src/main/java/com/helix/tasks/TasksService.kt
package com.helix.tasks

import com.helix.gateway.GatewaySession
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlin.coroutines.suspendCancellableCoroutine

class TasksService(private val gateway: GatewaySession) {
    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    // MARK: - Task Operations

    suspend fun getTasks(
        boardId: String,
        limit: Int = 50,
        offset: Int = 0,
    ): List<Task> = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "tasks.list",
                params = mapOf(
                    "board_id" to boardId,
                    "limit" to limit,
                    "offset" to offset,
                ),
            ) { response ->
                try {
                    val tasks = (response as? JsonElement)?.jsonArray?.map {
                        json.decodeFromJsonElement(Task.serializer(), it)
                    } ?: emptyList()
                    continuation.resume(tasks)
                } catch (e: Exception) {
                    continuation.resumeWithException(TaskError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun getTask(id: String): Task = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "tasks.get",
                params = mapOf("id" to id),
            ) { response ->
                try {
                    val task = json.decodeFromJsonElement(Task.serializer(), response as JsonElement)
                    continuation.resume(task)
                } catch (e: Exception) {
                    continuation.resumeWithException(TaskError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun createTask(
        boardId: String,
        title: String,
        description: String? = null,
        priority: TaskPriority = TaskPriority.Medium,
        dueDate: String? = null,
        estimatedHours: Double? = null,
        tags: List<String> = emptyList(),
    ): Task = suspendCancellableCoroutine { continuation ->
        try {
            val params = mutableMapOf<String, Any>(
                "board_id" to boardId,
                "title" to title,
                "priority" to priority.toString(),
                "tags" to tags,
            )

            description?.let { params["description"] = it }
            dueDate?.let { params["due_date"] = it }
            estimatedHours?.let { params["estimated_hours"] = it }

            gateway.request(
                method = "tasks.create",
                params = params,
            ) { response ->
                try {
                    val task = json.decodeFromJsonElement(Task.serializer(), response as JsonElement)
                    continuation.resume(task)
                } catch (e: Exception) {
                    continuation.resumeWithException(TaskError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun updateTask(
        id: String,
        title: String? = null,
        status: TaskStatus? = null,
        priority: TaskPriority? = null,
        dueDate: String? = null,
    ): Task = suspendCancellableCoroutine { continuation ->
        try {
            val params = mutableMapOf<String, Any>("id" to id)

            title?.let { params["title"] = it }
            status?.let { params["status"] = it.toString() }
            priority?.let { params["priority"] = it.toString() }
            dueDate?.let { params["due_date"] = it }

            gateway.request(
                method = "tasks.update",
                params = params,
            ) { response ->
                try {
                    val task = json.decodeFromJsonElement(Task.serializer(), response as JsonElement)
                    continuation.resume(task)
                } catch (e: Exception) {
                    continuation.resumeWithException(TaskError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun deleteTask(id: String): Unit = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "tasks.delete",
                params = mapOf("id" to id),
            ) { _ ->
                continuation.resume(Unit)
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    // MARK: - Subtask Operations

    suspend fun getSubtasks(taskId: String): List<Subtask> =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.subtasks.list",
                    params = mapOf("task_id" to taskId),
                ) { response ->
                    try {
                        val subtasks = (response as? JsonElement)?.jsonArray?.map {
                            json.decodeFromJsonElement(Subtask.serializer(), it)
                        } ?: emptyList()
                        continuation.resume(subtasks)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun createSubtask(taskId: String, title: String): Subtask =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.subtasks.create",
                    params = mapOf("task_id" to taskId, "title" to title),
                ) { response ->
                    try {
                        val subtask = json.decodeFromJsonElement(
                            Subtask.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(subtask)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun toggleSubtask(id: String, isCompleted: Boolean): Subtask =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.subtasks.toggle",
                    params = mapOf("id" to id, "is_completed" to isCompleted),
                ) { response ->
                    try {
                        val subtask = json.decodeFromJsonElement(
                            Subtask.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(subtask)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    // MARK: - Board Operations

    suspend fun getBoards(): List<KanbanBoard> = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "tasks.boards.list",
                params = emptyMap(),
            ) { response ->
                try {
                    val boards = (response as? JsonElement)?.jsonArray?.map {
                        json.decodeFromJsonElement(KanbanBoard.serializer(), it)
                    } ?: emptyList()
                    continuation.resume(boards)
                } catch (e: Exception) {
                    continuation.resumeWithException(TaskError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    // MARK: - Search

    suspend fun searchTasks(filter: TaskSearchFilter, limit: Int = 50): List<Task> =
        suspendCancellableCoroutine { continuation ->
            try {
                val params = mutableMapOf<String, Any>(
                    "limit" to limit,
                )

                if (filter.query.isNotEmpty()) params["query"] = filter.query
                if (filter.status.isNotEmpty()) params["status"] =
                    filter.status.map { it.toString() }
                if (filter.priority.isNotEmpty()) params["priority"] =
                    filter.priority.map { it.toString() }
                filter.assigneeId?.let { params["assignee_id"] = it }
                filter.boardId?.let { params["board_id"] = it }
                if (filter.tags.isNotEmpty()) params["tags"] = filter.tags
                filter.hasSubtasks?.let { params["has_subtasks"] = it }
                filter.isOverdue?.let { params["is_overdue"] = it }

                gateway.request(
                    method = "tasks.search",
                    params = params,
                ) { response ->
                    try {
                        val tasks = (response as? JsonElement)?.jsonArray?.map {
                            json.decodeFromJsonElement(Task.serializer(), it)
                        } ?: emptyList()
                        continuation.resume(tasks)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    // MARK: - Analytics

    suspend fun getAnalytics(boardId: String): TaskAnalytics =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.analytics",
                    params = mapOf("board_id" to boardId),
                ) { response ->
                    try {
                        val analytics = json.decodeFromJsonElement(
                            TaskAnalytics.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(analytics)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    // MARK: - Time Tracking

    suspend fun startTimeEntry(taskId: String): TimeEntry =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.timetracking.start",
                    params = mapOf("task_id" to taskId),
                ) { response ->
                    try {
                        val entry = json.decodeFromJsonElement(
                            TimeEntry.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(entry)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun stopTimeEntry(taskId: String): TimeEntry =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "tasks.timetracking.stop",
                    params = mapOf("task_id" to taskId),
                ) { response ->
                    try {
                        val entry = json.decodeFromJsonElement(
                            TimeEntry.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(entry)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    // MARK: - Focus Time

    suspend fun getFocusTimeSettings(): FocusTime = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "tasks.focustime.get",
                params = emptyMap(),
            ) { response ->
                try {
                    val settings = json.decodeFromJsonElement(
                        FocusTime.serializer(),
                        response as JsonElement
                    )
                    continuation.resume(settings)
                } catch (e: Exception) {
                    continuation.resumeWithException(
                        TaskError.DecodingError(e.message ?: "Unknown")
                    )
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun updateFocusTimeSettings(settings: FocusTime): FocusTime =
        suspendCancellableCoroutine { continuation ->
            try {
                val params = mapOf(
                    "daily_goal_minutes" to settings.dailyGoalMinutes,
                    "start_hour" to settings.startHour,
                    "end_hour" to settings.endHour,
                    "enable_notifications" to settings.enableNotifications,
                    "blocked_apps_url_patterns" to settings.blockedAppsUrlPatterns,
                )

                gateway.request(
                    method = "tasks.focustime.update",
                    params = params,
                ) { response ->
                    try {
                        val updated = json.decodeFromJsonElement(
                            FocusTime.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(updated)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            TaskError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }
}
